[FIX] l10n_it_edi: use date instead of invoice_date on self invoices

micheledic opened the pull request:
When generating the self-invoice XML, the system incorrectly uses the invoice_date field, which reflects the supplier’s invoice date. However, for self-invoices in Italy, the date field (i.e., the accounting date) should be used instead.

According to Italian regulations, self-invoices must be issued within 15 days of the transaction, and the invoice date must fall within the month the invoice is received. Using the supplier’s invoice date (e.g., April 2025) when the invoice is actually received in May 2025 results in non-compliance.

References
Official: https://www.agenziaentrate.gov.it/portale/documents/20143/451259/Guida_compilazione-FE-Esterometro-V_1.9_2024-03-05.pdf/67fe4c2d-1174-e8de-f1ee-cea77b7f5203 , page 14 
Extra: https://www.fiscoetasse.com/approfondimenti/16247-reverse-charge-interno-e-reverse-charge-esterno.html , entry 5

![image](https://github.com/user-attachments/assets/e593f463-a13a-460e-b125-9fbaaa674366)

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/210342.png)](https://mergebot.odoo.com/odoo/odoo/pull/210342)

Megaaaaaa commented on addons/l10n_it_edi/data/invoice_it_template.xml:
```diff
@@ -132,7 +132,7 @@
                 <TipoDocumento t-out="document_type"/>
                 <Divisa t-out="currency.name"/>
-                <Data t-out="format_date(record.invoice_date)"/>
+                <Data t-out="format_date(record.invoice_date if not is_self_invoice else record.date)"/>
```
Just for aesthetics, it'd be nicer to not use `not` :smile:
```suggestion
                <Data t-out="format_date(record.date if is_self_invoice else record.invoice_date)"/>
```

Megaaaaaa commented on addons/l10n_it_edi/tests/test_edi_reverse_charge.py:
```diff
@@ -180,6 +180,7 @@ def test_bill_reverse_charge_and_refund(self):
             ("Product B, taxed 4%", self.product_b, self.purchase_tax_4p),
             ("Product A", self.product_a, self.purchase_tax_22p),
         ]])
+        bills.write({'date': '2022-04-01'})
```
Can't you put all those `.write()`'s values directly in the `.create()` methods' dictionnaries ?

micheledic replied:
done

Megaaaaaa reviewed (commented):
Hello :wave:

Thank you for your fix!

I was wondering if you could explain me a few things so I can make sure I totally understand what's happening here.

1. Could you provide in the commit and PR message a link to the documentation which justifies the changes you apply ? It seems to be a legal issue so I guess this must come from the govmnt ?

2. Indirectly linked to 1, in the two templates of the file you are editing, there are multiple occurences of `invoice_date`being used but you only change one of those occurences to use `date`instead depending on if the move from which the date is used is an invoice or not. Is it meant ? Shouldn't your logic be applied to all dates coming from potential invoices ? (reconciled, reversed, downpayment, lines date) :thinking: 

3. You modify the tests to your adapt to the new behavior but you only have tests that now use the `record.date`. None still use `record.invoice_date` if I'm not mistaken. Would you mind adding a new simple test to keep testing that the `invoice_date` is also used when we need it to ? Or tell me where I've missed it if it already exists :smile:

Thank you a lot for your time and, in advance, for your answer ! :pray:

micheledic commented:
> Hello 👋
> 
> Thank you for your fix!
> 
> I was wondering if you could explain me a few things so I can make sure I totally understand what's happening here.
> 
> 1. Could you provide in the commit and PR message a link to the documentation which justifies the changes you apply ? It seems to be a legal issue so I guess this must come from the govmnt ?
> 2. Indirectly linked to 1, in the two templates of the file you are editing, there are multiple occurences of `invoice_date`being used but you only change one of those occurences to use `date`instead depending on if the move from which the date is used is an invoice or not. Is it meant ? Shouldn't your logic be applied to all dates coming from potential invoices ? (reconciled, reversed, downpayment, lines date) 🤔
> 3. You modify the tests to your adapt to the new behavior but you only have tests that now use the `record.date`. None still use `record.invoice_date` if I'm not mistaken. Would you mind adding a new simple test to keep testing that the `invoice_date` is also used when we need it to ? Or tell me where I've missed it if it already exists 😄
> 
> Thank you a lot for your time and, in advance, for your answer ! 🙏

Hello,
1) added the documentation about the date
2) changed the occurrencies to .date if self invoice
3) there are tests on "invoice_date" on other cases . for example 
https://github.com/odoo/odoo/blob/6b85b36f036638357e81427bcc393ee8394eb927/addons/l10n_it_edi/tests/test_edi_export.py#L104-L132

Megaaaaaa commented:
Thank you for the clarification and the changes, looks good. We're nearly there, there's just one last thing I'd like to check before merging and also I realized that this could be fixed in 16.0 too would you mind retargetting the PR please ?

lordkrandel commented:
It's a good idea to do it for `Odoo 17`.
Also, I have to remark that for our stable policy and recent internal remarks, `Odoo 16` bugfixes must be limited to critical issues and tracebacks, as it nears 3 years since release. This will avoid excessive forward ports, conflicts, issues, as shortly after September, support for the version will be dropped, with the release of `Odoo 19`

micheledic commented:
> Legal issues are critical and should be fixed in all supported versions in my opinion. If we cause people to be outside the law because something isn't done properly in our software, we are not doing a good job. Until 16.0 is not supported, we should do everything we can to comply with the law, an additionnal forward port shouldn't be an issue. If you don't agree, don't hesitate to say it, I'd like to understand why if I'm overreacting about this.

Hello, so can we do a separate forward port ?
As i said if i target to 16.0, i have to remove some changes i already did for the 17.0 because the code and tests are changed since 17 and i have not the permission on odoo-dev organization (where the code is pushed by robodoo).
It will be easy to forward port as separate PR in the next future to the 16.0

Megaaaaaa commented:
Yes please open a second PR for the backport in 16.0 and you can ping me on it, I'll handle it in top priority. Thank you again for your contribution! :pray:

robodoo r+
